File: javascript/bob/bob.js
//
// This is only a SKELETON file for the 'Bob' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const hey = (msg) => {
  if (isShouting(msg) && isQuestion(msg)) return 'Calm down, I know what I\'m doing!';
  if (isQuestion(msg)) return 'Sure.';
  if (isShouting(msg)) return 'Whoa, chill out!';
  if (isSilence(msg)) return 'Fine. Be that way!';
  return 'Whatever.';
};

function isSilence(msg) {
  return msg.match(/^\s*$/);
}

function isShouting(msg) {
  return msg.match(/[A-Z]/) && !msg.match(/[a-z]/);
}

function isQuestion(msg) {
  return msg.match(/\?\s*$/);
}

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      silence?(input) ->
        "Fine. Be that way!"
      question_in_capitals?(input) ->
        "Calm down, I know what I'm doing!"
      question?(input) ->
        "Sure."
      shouting?(input) ->
        "Whoa, chill out!"
      true ->
        "Whatever."
    end
  end

  defp silence?(input) do
    String.trim(input) |> String.length == 0
  end

  defp shouting?(input) do
    capitals?(input)
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end

  defp question_in_capitals?(input) do
    question?(input) && capitals?(input)
  end

  defp capitals?(input) do
    String.upcase(input) == input && String.downcase(input) != input
  end
end
